kpops/utils/dict_ops.py

```diff
@@ -0,0 +1,129 @@
+) -> dict[str, Any]:
+    """Add all nested key-value pairs to the top level of the mapping
+
+    Does not remove any fields, only duplicates existing ones and moves them up
+    with a corresponding prefix.
+    Hence, output is not exactly flat.
+
+    :param nested_mapping: Nested mapping that is to be `inflated`
+    :type nested_mapping: Mapping[str, any]
+    :param prefix: Prefix that will be applied to all top-level keys in the output., defaults to None
+    :type prefix: str, optional
+    :param separator: Separator between the prefix and the keys, defaults to "_"
+    :type separator: str, optional
+    :returns: "Flattened" mapping in the form of dict
+    :rtype: dict[str, Any]
+    """
+    if not isinstance(nested_mapping, Mapping):
+        raise TypeError("Argument nested_mapping is not a Mapping")
+    top = {}
+    for key, value in nested_mapping.items():
+        if not isinstance(key, str):
+            raise TypeError(f"Argument nested_mapping contains a non-str key: {key}")
+        if prefix:
+            key = prefix + separator + key
+        top[key] = value
+        if isinstance(value, Mapping):
+            nested_mapping = inflate_mapping(value, key)
+            top = update_nested_pair(top, nested_mapping)
+    return top
+
+
+def generate_substitution(
+    input: dict,
+    prefix: str | None = None,
+    existing_substitution: dict | None = None,
+) -> dict:
+    """Generate a complete substitution dict from a BaseModel
+
+    Finds all attributes that belong to a model and expands them to create
+    a dict containing each variable name and value to substitute with.
+
+    :param input: Dict from which to generate the substitution
+    :type input: dict
+    :param prefix: Prefix the preceeds all substitution variables, defaults to None
+    :type prefix: str, optional
+    :param substitution: existing substitution to include
+    :type substitution: dict
+    :returns: Substitution dict of all variables related to the model.
+    :rtype: dict
+    """
+    existing_substitution = is_dict(
+        existing_substitution,
+        "Argument ``existing_substitution`` must be of type dict or None",
+    )
+    return update_nested(existing_substitution, inflate_mapping(input, prefix))
+
+
+def is_dict(d, ex_msg: str = "Argument must be of type dict or None") -> dict:
+    """Check whether the input is a ``dict``
+
+    If it is ``None``, an empty dict is returned. Else a ``TypeError`` is raised.
+
+    :param d: Input to be checked
+    :type d: Any
+    :param ex_msg: Exception message, defaults to "Argument must be of type dict"
+    :type ex_msg: str, optional
+    :raises TypeError: Argument must be of type dict or None
+    :returns: unmodified ``d`` or empty ``dict`` if d is ``None``
+    """
+    if d is None:
+        return {}
+    elif isinstance(d, dict):
+        return d
+    else:
+        raise TypeError(ex_msg)
```

Comment:
why would it not be `dict | None`? could you give an example for such a scenario?

Reply:
See [here](https://github.com/bakdata/kpops/pull/198/files#diff-a7e11ead0d843c3b75b7427119e68b95eb41970f919eff2cf96e1111ff485822R86-R109)

I [extracted](https://github.com/bakdata/kpops/pull/198/commits/a0e6ffe1bcd027aba99fe8cdf921f0f2afe1d5b9#diff-a7e11ead0d843c3b75b7427119e68b95eb41970f919eff2cf96e1111ff485822L102-L109) the check out of `generate_substitution()` following [this discussion](https://github.com/bakdata/kpops/pull/198/files/1823e6a1be8961046c03af13076be028d09396f4#r1211217053)

Reply:
yeah, I was wondering because there it is defined as `existing_substitution: dict | None = None`. so then how could it be of some other type?

Reply:
It is a bit of a defensive approach. Tbh I prefer to return the check inside of the function that sits right above it

This function seems a bit too specific to be on its own like that

Reply:
I simplified it a lot now